Plugins/BuildServerIntegration/AzureDevOpsIntegration/ApiClient.cs

```diff
@@ -109,26 +110,45 @@ public async Task<string> GetBuildDefinitionNameFromIdAsync(int buildId)
             return build.Definition.Name;
         }
 
-        public async Task<IEnumerable<Build>> QueryBuildsAsync(string buildDefinitionsToQuery, DateTime? sinceDate, bool? running)
+        public async Task<IEnumerable<Build>> QueryFinishedBuildsAsync(string buildDefinitionsToQuery, DateTime? sinceDate)
         {
-            if (buildDefinitionsToQuery == null)
+            string queryUrl = QueryForBuildStatus(buildDefinitionsToQuery, "completed");
+            queryUrl += sinceDate.HasValue
+                ? $"&minTime={sinceDate.Value.ToUniversalTime():s}&api-version=4.1"
+                : "&api-version=2.0";
+
+            var finishedBuilds = (await HttpGetAsync<ListWrapper<Build>>(queryUrl)).Value;
+            return finishedBuilds;
+        }
+
+        public async Task<IEnumerable<Build>> QueryRunningBuildsAsync(string buildDefinitionsToQuery)
+        {
+            string queryUrl = QueryForBuildStatus(buildDefinitionsToQuery, "cancelling,inProgress,none,notStarted,postponed") + "&api-version=2.0";
+
+            var runningBuilds = (await HttpGetAsync<ListWrapper<Build>>(queryUrl)).Value;
+
+            if (runningBuilds.Count < 2)
             {
-                return Enumerable.Empty<Build>();
+                return runningBuilds;
             }
 
-            var builds = (await HttpGetAsync<ListWrapper<Build>>($"build/builds?api-version=2.0&definitions={buildDefinitionsToQuery}")).Value;
+            var byCommitBuilds = runningBuilds.GroupBy(b => b.SourceVersion);
+            runningBuilds = new List<Build>();
 
-            if (!running.HasValue || running.Value)
+            // Select best build to display (the first started or if none, one that will start)
+            foreach (var commitBuilds in byCommitBuilds)
             {
-                return builds
-                    .Where(b => !running.HasValue || running.Value == b.IsInProgress)
-                    .Where(b => !sinceDate.HasValue || b.StartTime >= sinceDate.Value.ToUniversalTime());
+                var builds = commitBuilds.OrderBy(b => b.StartTime).ToArray();
```

Comment:
Why do we need to materialise the query to take a top element form it? Feels wasteful, especially for a plugin that could be constantly running for a life time of the app (which can be quite long) and with multiple instances of the app open at the same time - we could be consuming more memory than necessary...

To me `.OrderBy(b => b.StartTime)` followed by `FirstOrDefault(b => b.StartTime.HasValue)` looks sus.
Could you please elaborate on what we  are doing here?

Can you please write some tests with some real response data? We need to restructure the code a little, e.g.
```cs
        public async Task<IEnumerable<Build>> QueryRunningBuildsAsync(string buildDefinitionsToQuery)
        {
            string queryUrl = QueryForBuildStatus(buildDefinitionsToQuery, "cancelling,inProgress,none,notStarted,postponed") + "&api-version=2.0";
            var runningBuilds = (await HttpGetAsync<ListWrapper<Build>>(queryUrl)).Value;
            return ParseResponse(runningBuilds);
        }
```

...and test the `ParseResponse` method.

Hint, to assert blobs of data ApprovalTests can be quite useful.

Reply:
>Could you please elaborate on what we are doing here?

You're right, that was not optimal. It was to reuse the ordered values after the `??` if no values was found but thinking about it again that was not necessary as we could pick the first value in the unsorted collection.

So fixed.

I will have a look at the other concern...